test/torchaudio_unittest/rnnt/autograd_impl.py

```diff
@@ -0,0 +1,148 @@
+        data = self.get_data(data_func, self.device)
+        pytorch_rnnt_loss = RNNTLoss(blank=data["blank"])
+        costs = pytorch_rnnt_loss(data["logits"],
+                                  data["targets"],
+                                  data["logit_lengths"],
+                                  data["target_lengths"])
+        loss = torch.sum(costs)
+        loss.backward()
+        return loss, data["logits"].grad
+
+    def assert_grad(
+            self,
+            loss: Callable[..., Tensor],
+            inputs: Tuple[torch.Tensor],
+            *,
+            enable_all_grad: bool = True,
+            nondet_tol: float = 0.0,
+    ):
+        # inputs_ = []
+        # for i in inputs:
+        #     if torch.is_tensor(i):
+        #         i = i.to(dtype=self.dtype, device=self.device)
+        #         if enable_all_grad:
+        #             i.requires_grad = True
+        #     inputs_.append(i)
+        assert gradcheck(loss, inputs, eps=1e-03, atol=1e-02, nondet_tol=nondet_tol)
+
+    @parameterized.expand([
+        (get_B1_T10_U3_D4_data, ),
+        (get_numpy_data_B2_T4_U3_D3, ),
+        (get_numpy_data_B1_T2_U3_D5, ),
+    ])
+    def test_RNNTLoss_gradcheck(self, data_func):
+        data = self.get_data(data_func, self.device)
+        inputs = (
+            data["logits"].to(self.dtype),
+            data["targets"],
+            data["logit_lengths"],
+            data["target_lengths"],
+        )
+        loss = RNNTLoss(blank=data["blank"])
+        
+        self.assert_grad(loss, inputs, enable_all_grad=False, nondet_tol=1e-10)
+
+    @parameterized.expand([
+        (get_B1_T10_U3_D4_data, ),
+        (get_numpy_data_B2_T4_U3_D3, ),
+        (get_numpy_data_B1_T2_U3_D5, ),
+    ])
+    def test_rnnt_loss_gradcheck(self, data_func):
+        data = self.get_data(data_func, self.device)
+        inputs = (
+            data["logits"].to(self.dtype),
+            data["targets"],
+            data["logit_lengths"],
+            data["target_lengths"],
+        )
+        loss = RNNTLoss(blank=data["blank"])
+        
+        self.assert_grad(loss, inputs, enable_all_grad=False, nondet_tol=1e-10)
+
+    @parameterized.expand([
+        (get_B1_T10_U3_D4_data, ),
+        (get_numpy_data_B2_T4_U3_D3, ),
+        (get_numpy_data_B1_T2_U3_D5, ),
+    ])
+    def test_np_transducer_gradcheck(self, data_func):
+        data = self.get_data(data_func, self.device)
+        inputs = (
+            data["logits"].to(self.dtype),
+            data["logit_lengths"],
+            data["target_lengths"],
+            data["targets"],
+        )
+        loss = NumpyTransducerLoss(blank=data["blank"])
+        
+        self.assert_grad(loss, inputs, enable_all_grad=False, nondet_tol=1e-10)
+
+    @parameterized.expand([
+        (get_B1_T10_U3_D4_data, ),
+        (get_numpy_data_B2_T4_U3_D3, ),
+        # (get_numpy_data_B1_T2_U3_D5, ),
+    ])
+    def test_warp_transducer_gradcheck(self, data_func):
```

Comment:
the open source warp-transducer passes gradcheck for `get_B1_T10_U3_D4_data` and `get_numpy_data_B2_T4_U3_D3`, and we have also confirmed that the gradients of the torchaudio and warp-transducer implementations match for both these data sets (comparing the .grad attribute after calling backwards). However, torchaudio is failing gradcheck due to `Backward is not reentrant, i.e., running backward with same input and grad_output multiple times gives different values, although analytical gradient matches numerical gradient.The tolerance for nondeterminism was 1e-10.`

@albanD are there any reasons you can think of for this behavior?